controlplane/kubeadm/internal/cluster.go

```diff
@@ -0,0 +1,389 @@
+		Log.Info("node", "node-name", node.Name)
+	}
+
+	tlsConfig, err := c.generateEtcdTLSClientBundle()
+	if err != nil {
+		return nil, err
+	}
+
+	response := map[string]error{}
+	for _, node := range controlPlaneNodes.Items {
+		name := node.Name
+		response[name] = nil
+		if node.Spec.ProviderID == "" {
+			response[name] = errors.New("empty provider ID")
+			continue
+		}
+
+		// Create the etcd client for the etcd Pod scheduled on the Node
+		etcdClient, err := c.getEtcdClientForNode(name, tlsConfig)
+		if err != nil {
+			response[name] = errors.New("failed to create etcd client")
+			continue
+		}
+
+		// List etcd members. This checks that the member is healthy, because the request goes through consensus.
+		members, err := etcdClient.Members(ctx)
+		if err != nil {
+			response[name] = errors.New("failed to list etcd members using etcd client")
+			continue
+		}
+		Log.Info("members for specific node", "node", name, "members", members)
+		member := etcdutil.MemberForName(members, name)
+
+		// Check that the member reports no alarms.
+		if len(member.Alarms) > 0 {
+			response[name] = errors.Errorf("etcd member reports alarms: %v", member.Alarms)
+			continue
+		}
+
+		// Check that the member belongs to the same cluster as all other members.
+		clusterID := member.ClusterID
+		if knownClusterID == 0 {
+			knownClusterID = clusterID
+		} else if knownClusterID != clusterID {
+			response[name] = errors.Errorf("etcd member has cluster ID %d, but all previously seen etcd members have cluster ID %d", clusterID, knownClusterID)
+			continue
+		}
+
+		// Check that the member list is stable.
+		// TODO: this isn't necessary but reduces the race condition as much as possible.
+		memberIDSet := etcdutil.MemberIDSet(members)
+		if knownMemberIDSet.Len() == 0 {
+			knownMemberIDSet = memberIDSet
+		} else {
+			unknownMembers := memberIDSet.Difference(knownMemberIDSet)
+			if unknownMembers.Len() > 0 {
+				response[name] = errors.Errorf("etcd member reports members IDs %v, but all previously seen etcd members reported member IDs %v", memberIDSet.UnsortedList(), knownMemberIDSet.UnsortedList())
+			}
+			continue
+		}
+	}
+
+	// Check that there is exactly one etcd member for every control plane machine.
+	// There should be no etcd members added "out of band.""
+	if len(controlPlaneNodes.Items) != len(knownMemberIDSet) {
+		return response, errors.Errorf("there are %d control plane nodes, but %d etcd members", len(controlPlaneNodes.Items), len(knownMemberIDSet))
+	}
+
+	return response, nil
+}
+
+// getEtcdClientForNode returns a client that talks directly to an etcd instance living on a particular node.
+func (c *cluster) getEtcdClientForNode(nodeName string, tlsConfig *tls.Config) (*etcd.Client, error) {
+	// This does not support external etcd.
+	p := proxy.Proxy{
+		Kind:         "pods",
+		Namespace:    "kube-system", // TODO, can etcd ever run in a different namespace?
+		ResourceName: fmt.Sprintf("etcd-%s", nodeName),
```

Comment:
nit: Might be worth noting this produces the right name because the etcd Pod has the name `etcd`, but because it is a Static Pod, the node name gets appended to its name.

Reply:
i'll leave a note

Reply:
actually i just pulled it into a fn so it's extremely clear